src/models/datastore.js

```diff
@@ -0,0 +1,261 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+let ModelFactory = require('../helpers/model-factory');
+let Connection = require('./connection');
+
+class DataStore {
+
+  constructor () {
+    this.storeName_ = null;
+  }
+
+  configure (opts) {
+    if (typeof opts === 'undefined') {
+      throw new Error('DataStore configuration not provided.');
+    }
+
+    this.storeName = opts.store || null;
+    ModelFactory.modelData = opts.models || null;
+  }
+
+  get storeName () {
+    return this.storeName_;
+  }
+
+  set storeName (storeName) {
+    this.storeName_ = storeName;
+  }
+
+  get models () {
+    return ModelFactory.modelData;
+  }
+
+  set models (modelData) {
+    ModelFactory.modelData = modelData;
+  }
+
+  _ensureValidPopulations (populations) {
+    // Assume missing.
+    if (typeof populations === 'undefined' || populations === null) {
+      return [];
+    }
+
+    // Assume string.
+    if (typeof populations === 'string') {
+      return [populations];
+    }
+
+    // Assume object.
+    if (typeof populations.length === 'undefined') {
+      return [populations];
+    }
+
+    // Assume array.
+    return populations;
+  }
+
+  action_ (type, cb) {
+    if (this.storeName === null) {
+      return Promise.reject(new Error('Database name must be set.'));
+    }
+
+    let connection;
+
+    return Connection
+
+        // Get a connection.
+        .getFromPool(this.storeName)
+
+        // Store it so it can be released later.
+        .then(conn => {
+          connection = conn;
+        })
+
+        // Call the callback.
+        .then(() => {
+          let model = ModelFactory.get(type);
+
+          if (model === null) {
+            return Promise.reject(new Error('No model found for ' + type));
+          }
+
+          return cb(model);
+        })
+
+        // Propagate the result
+        .then(result => {
+          Connection.returnToPool(connection);
+          return result;
+        }).catch(err => {
+          Connection.returnToPool(connection);
+          throw err;
+        });
+  }
+
+  put (type, data, id) {
+    // If an array bounce it down to individual requests.
+    if (Array.isArray(data)) {
+      // Check that every item in the array is at least an object.
+      if (!data.every(item => typeof item === 'object')) {
+        return Promise.reject(
+          new Error('put() given array of entries which are not all objects.')
+        );
+      }
+
+      // Remap to an array of promises for each item.
+      return Promise.all(data.map((item) => this.put(type, item)));
+    }
+
+    // Make sure that an object is passed.
+    if (typeof data !== 'object') {
+      return Promise.reject(new Error('put() requires an object.'));
+    }
+
+    // Check if the data is an existing model. If so, fail.
+    if (typeof data._id === 'object' &&
+        data._id.constructor.name === 'ObjectID') {
```

Comment:
ObjectID is Mongoose, and if you pass _that_ rather than an object then Mongoose would fail, so I'm catching the case where you pass an existing model rather than a plain object.